solana/solana-ibc/programs/solana-ibc/src/tests.rs

```diff
@@ -54,6 +73,66 @@ macro_rules! make_message {
     }}
 }
 
+pub struct DeliverWithRemainingAccounts {
+    sender: Pubkey,
+    storage: Pubkey,
+    trie: Pubkey,
+    packets: Pubkey,
+    chain: Pubkey,
+    system_program: Pubkey,
+    remaining_accounts: Vec<Pubkey>,
+}
+
+impl ToAccountMetas for DeliverWithRemainingAccounts {
+    fn to_account_metas(
+        &self,
+        _is_signer: Option<bool>,
+    ) -> Vec<anchor_lang::prelude::AccountMeta> {
+        let mut accounts = Vec::new();
+        accounts.push(AccountMeta {
+            pubkey: self.sender,
+            is_signer: true,
+            is_writable: true,
+        });
+        accounts.push(AccountMeta {
+            pubkey: self.storage,
+            is_signer: false,
+            is_writable: true,
+        });
+        accounts.push(AccountMeta {
+            pubkey: self.trie,
+            is_signer: false,
+            is_writable: true,
+        });
+        accounts.push(AccountMeta {
+            pubkey: self.packets,
+            is_signer: false,
+            is_writable: true,
+        });
+        accounts.push(AccountMeta {
+            pubkey: self.chain,
+            is_signer: false,
+            is_writable: true,
+        });
+        accounts.push(AccountMeta {
+            pubkey: self.system_program,
+            is_signer: false,
+            is_writable: false,
+        });
+        
+
+        self.remaining_accounts.iter().for_each(|&account| {
+            accounts.push(AccountMeta {
+                pubkey: account,
+                is_signer: false,
+                is_writable: false,
+            })
+        });
+
+        accounts
```

Comment:
Isn’t there some Anchor derive that can do this for us?  Can’t we
`#[derive(Account)]` or something on the struct?

---

As per existing code, a few small things for minor optimisations.

Firstly, you can do `[...].to_vec()` to avoid the repetition of
`accounts.push`.  I think it ends up looking cleaner and it also
potentially works better with minimising allocations (though to be
honest I dunno what’s vector’s first allocation size so this may be
a moot point).

Secondly, I would suggest avoiding `for_each` function.  It’s nearly
always better to use an explicit loop.  `for_each` rarely gives you
anything and it introduces a context where keywords change meaning
(`return` is now `break`, `break` is not valid) making the code harder
to understand.

And since you’re pushing to a vector, it’s probably better to use
`Vec::extend` anyway.

But overall, if you really care about micro optimisations, this whole
thing can be written with iterator chaining and a single collect
I believe:

```suggestion
        let accounts = [
            AccountMeta {
                pubkey: self.sender,
                is_signer: true,
                is_writable: true,
            },
            AccountMeta {
                pubkey: self.storage,
                is_signer: false,
                is_writable: true,
            },
            AccountMeta {
                pubkey: self.trie,
                is_signer: false,
                is_writable: true,
            },
            AccountMeta {
                pubkey: self.packets,
                is_signer: false,
                is_writable: true,
            },
            AccountMeta {
                pubkey: self.chain,
                is_signer: false,
                is_writable: true,
            },
            AccountMeta {
                pubkey: self.system_program,
                is_signer: false,
                is_writable: false,
            },
        ];

        let remaining = self
            .remaining_accounts
            .iter()
            .map(|&account| AccountMeta {
                pubkey: account,
                is_signer: false,
                is_writable: false,
            });

        accounts.into_iter().chain(remaining).collect::<Vec<_>>()
```

Reply:
right right, that makes total sense.

Reply:
> Isn’t there some Anchor derive that can do this for us? Can’t we
#[derive(Account)] or something on the struct?

We have to use our own struct since we need to pass the remaining accounts. The anchor derive doesnt enable us to do that.